### ISSUE TITLE
Bad performance when running homemade OS in qemu

### ISSUE THREAD

**Issue by perlun** [tsl0922/ttyd#137]
Hi,

Thanks for an interesting project. Trying it out since I'm looking for a way to publish the [chaos](http://www.chaosdev.io) operating system on the web, to make it easy for people who want to play around with it.

However, the performance is not very great in my use case which makes things less than ideal.

This is the command line I'm trying:

```shell
$ ttyd -m10 docker run -it chaosdev/chaos
```

(For the curious, we build a Docker image for each push to `master` using [this Dockerfile](https://github.com/chaos4ever/chaos/blob/master/Dockerfile))

When connecting from "Firefox on Windows" -> "`ttyd` inside a Linux VM", the experience is unfortunately quite sluggish and a bit sub-par. Typing characters on the keyboard shows a significant input lag. However, when swapping Docker container to instead use a plain `debian` image for example, it works great and the performance is awesome:

```shell
$ ttyd -m10 docker run -it debian
```

The `chaos` image uses a custom background, so there is a bit of painting to do for each scroll/etc, but even general typing at the command (when not very much scrolling is done in the underlying console system in the OS) is very sluggish. It's a huge difference from running the same container right in a Linux terminal - it is much more responsive.

Even the GRUB bootloader (used by chaos before the actual OS is loaded) is quite slow/lagging in terms of keyboard input, and it doesn't use any custom background or anything so maybe it's just the QEMU curses support that generates an extreme amount of terminal output that the Xterm.js layer needs to parse or...? I really don't know; any ideas would be extremely helpful here.

(Similar to #136, but perhaps a different root cause.)

## Additional info: inspecting network traffic when running the app over SSH instead of ttyd

I tried running the same Docker container via `ssh localhost`, and the performance is noticeably better than via ttyd. Here is a tcpdump of a single keypress - 36 bytes for the keypress, and 84 bytes for the response. So the payloads do not seem to be very large; something else is likely the root cause here.

```
13:58:40.212733 IP6 localhost.48122 > localhost.ssh: Flags [P.], seq 108:144, ack 453, win 3636, options [nop,nop,TS val 376876817 ecr 376873028], length 36
13:58:40.212778 IP6 localhost.ssh > localhost.48122: Flags [.], ack 144, win 1430, options [nop,nop,TS val 376876817 ecr 376876817], length 0

13:58:40.258684 IP6 localhost.ssh > localhost.48122: Flags [P.], seq 453:537, ack 144, win 1430, options [nop,nop,TS val 376876863 ecr 376876817], length 84
13:58:40.258739 IP6 localhost.48122 > localhost.ssh: Flags [.], ack 537, win 3636, options [nop,nop,TS val 376876863 ecr 376876863], length 0
```

**Comment by tsl0922** [comment]
This should be an upstream issue. I've upgraded xterm.js to the `3.8.0` version, it works better on my machine, try the latest `master` branch again.

**Comment by perlun** [comment]
Thanks @tsl0922, works better now. Closing issue. 👍

**Comment by perlun** [comment]
(For reference, the fix that helped is 4d5a099c5865e83f4e2e843765b077599c9b3047)
